### PR TITLE
feat: add copy to clipboard on json data

### PR DESCRIPTION
**Pull request by kshitij-k-osmosys**
### Pre-requisites

- [x] I have gone through the Contributing guidelines for [Submitting a Pull Request (PR)](https://github.com/OsmosysSoftware/osmo-x/blob/main/CONTRIBUTING.md#-submitting-a-pull-request-pr) and ensured that this is not a duplicate PR.
- [ ] I have performed unit testing for the new feature added or updated to ensure the new features added are working as expected.
- [x] I have performed preliminary testing to ensure that any existing features are not impacted and any new features are working as expected as a whole.

### PR Details

PR details have been updated as per the given format (see below)

- [x] PR title adheres to the format specified in guidelines (e.g., `feat: add admin login page`)
- [x] Description has been added
- [x] Related changes have been added (optional)
- [x] Screenshots have been added (optional, may include unit testing screenshots as well)
- [ ] Pending actions have been added (optional)
- [ ] Any other additional notes have been added (optional)

### Additional Information

- [x] Appropriate label(s) have been added (`ready for review` should be added if the PR is ready to be reviewed)
- [x] Assignee(s) and reviewer(s) have been added (optional)

_Note: Reviewer should ensure that the checklist and description have been populated and followed correctly, and the PR should be merged only after resolving all conversations and verifying that CI checks pass._

---

**Description:**

Add a button to Dialog box to copy JSON Data

**Related changes:**

- Check out angular cdk from angular material that we can use for copying data: https://material.angular.io/cdk/categories
- Install the v17 angular cdk on portal: https://www.npmjs.com/package/@angular/cdk/v/17.3.10
- Create a new button on json dialog box to copy
- Use JSON.stringify() to convert json to string before copying
- Add a primeng toast info notification when we successfully copy data
- Add error handling
- Tested the changes on local
- Tested the changes on docker

**Screenshots:**

- Sample video illustrating the copy feature
https://github.com/user-attachments/assets/c98dc1a0-5943-4737-aef5-dcf6946f9f5a

- Copy button in `View Data`
image

- Copy button in `View Result` with success toast
image

**Pending actions:**

NA

**Additional notes:**

NA

## Summary by CodeRabbit

- **New Features**
	- Added a "Copy JSON to Clipboard" button in the JSON dialog for easier data handling.
	- Integrated clipboard functionality and user feedback notifications for copy actions.
- **Enhancements**
	- Included Angular Component Development Kit (CDK) for improved UI components and application functionality.
- **Bug Fixes**
	- Improved test coverage for the JSON dialog component to ensure proper functionality of the copy feature and error handling.

### CODE REVIEW

File: apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.ts [hunk 1]
```diff
@@ -12,8 +14,33 @@ export class JsonDialogComponent {
...
+      });
+    } catch (error) {
+      this.messageService.add({
+        key: 'tst',
+        severity: 'error',
+        summary: 'Error',
+        detail: `There was an error copying JSON data. Reason: ${error.message}`,
+      });
+    }
+  }
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.ts:45]: how do I do that, in which file do I add this

[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.ts:45]: Added in a11820d6ef7e5331f2a7ad27594dac0e78168178

File: apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.spec.ts [hunk 2]
```diff
@@ -1,21 +1,56 @@
...
+import { TestBed } from '@angular/core/testing';
+import { Clipboard } from '@angular/cdk/clipboard';
+import { MessageService } from 'primeng/api';
 import { JsonDialogComponent } from './json-dialog.component';
 
 describe('JsonDialogComponent', () => {
   let component: JsonDialogComponent;
-  let fixture: ComponentFixture<JsonDialogComponent>;
+  let clipboardSpy: jasmine.SpyObj<Clipboard>;
+  let messageServiceSpy: jasmine.SpyObj<MessageService>;
```
[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.spec.ts:9]: give me an optimised .spec file for it

[the remark this answers is not in the document]
Comment (kshitij-k-osmosys) [reply apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.spec.ts:9]: fixed 1ef621defb007f62802b9e6668b2e41939274991

**Comment by xixas** [comment]
@kshitij-k-osmosys  Add screenshots/video

File: apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.spec.ts [hunk 3]
```diff
@@ -1,21 +1,53 @@
...
 
   beforeEach(() => {
     TestBed.configureTestingModule({
       declarations: [JsonDialogComponent],
+      providers: [
```
Comment (xixas) [inline apps/portal/src/app/views/notifications/json-dialog/json-dialog.component.spec.ts:14]: No need to write tests right now. Don't remove though

**Review by xixas — CHANGES_REQUESTED** [review]

**Review by xixas — APPROVED** [review]

**Pull request merged**